plenum/server/consensus/consensus_data_provider.py

```diff
@@ -48,22 +60,85 @@ def is_primary(self) -> bool:
     def preprepared(self) -> List[PrePrepare]:
         """
         List of PrePrepare messages, for which quorum of Prepare messages is not reached yet
-        TODO: Make it simple field?
         """
-        return []
+        return self._preprepared
 
     @property
     def prepared(self) -> List[PrePrepare]:
         """
         List of PrePrepare messages, for which quorum of Prepare messages is reached
-        TODO: Make it simple field?
         """
-        return []
+        return self._prepared
 
     @property
     def stable_checkpoint(self) -> int:
-        return 0
+        return self._stable_checkpoint
 
     @property
     def checkpoints(self) -> List[Checkpoint]:
-        return []
+        """
+        List of Checkpoints, which has been sent, but not stabilized yet
+        """
+        return self._checkpoints.values()
+
+    def preprepare_batch(self, pp: PrePrepare):
+        """
+        After pp had validated, it placed into _preprepared list
+        """
+        self._preprepared.append(pp)
+
+    def prepare_batch(self, pp: PrePrepare):
+        """
+        After prepared certificate for pp had collected,
+        it removed from _preprepared and placed into _prepared list
+        """
+        if pp is None:
+            raise LogicError('pp does not exist')
+        if pp not in self._preprepared:
+            raise LogicError('Unprepared pp must be stored in preprepared')
+        self._preprepared.remove(pp)
+        self._prepared.append(pp)
+
+    def free_batch(self, pp):
+        """
+        When 3pc batch had committed, rejected or catchuped till, it removed from _prepared list
+        """
+        if pp is None:
+            raise LogicError('pp does not exist')
+        if pp in self._preprepared:
```

Comment:
What if the batch is just preprepared and catchup is goint to be started? In this case we need to remove it from `_preprepared`. so this LogicError looks incorrect.

Reply:
right, fixed